Louis/notifier

ldelossa opened the pull request:
PR implementing ClairV4 notifications.

ldelossa committed:
notifier persistence layer implemented

Adds a store implementation and all methods necessary to create,
persist, and deliver notifications

Signed-off-by: ldelossa <ldelossa@redhat.com>

Commit:
implement poller and processor

Implements a poller abstraction which ingresses an event when a new
update operation is detected.
Implements processors which respond to this event and conditionally
create and persist notifications if appropriate.

Signed-off-by: ldelossa <ldelossa@redhat.com>

Commit:
implement delivery

Implements Delivery control structure which handles the logic of
delivering notifications.
Implements STOMP, AMQP and Webhook Deliverers which handle the infra-structure
specific details of notification delivery

Signed-off-by: ldelossa <ldelossa@redhat.com>

Commit:
implement service

Implements the service interface and local service wrapper around the
Notifier's facilities.

Signed-off-by: ldelossa <ldelossa@redhat.com>

Commit:
http and service plumbing

Small refactor to expand differ client into a full Matcher
implementation.
Introduce notifications http endpoint.
Wire notifier into the initialization control structure.
Introduce notifier into local dev environment.

Signed-off-by: ldelossa <ldelossa@redhat.com>

Allda commented on config.yaml.sample:
```diff
@@ -36,6 +36,27 @@ matcher:
+        type: "direct"
+        durable: true
+        auto_delete: false
+    uri: "amqp://user:pass@host:10000/vhost"
```
Does this support multiple brokers with failover?

ldelossa replied:
Wouldnt failover be handles by a proxy service ? If multiple urls is provided, what would be the correct one to connect to?

Allda commented on local-dev/clair/config.yaml:
```diff
@@ -12,6 +12,26 @@ matcher:
+  # webhook:
+  #   target: "http://webhook/"
+  #   callback: "http://clair-notifier/api/v1/notifications"
+  amqp:
```
amqp plugin should support tls auth

ldelossa replied:
Ok i think i can add this in.

Allda commented on notifier/amqp/deliverer.go:
```diff
@@ -0,0 +1,77 @@
+		Body:        b,
+	}
+	err = ch.Publish(
+		d.conf.Exchange.Name,
```
Is this a queue/topic where message is sent?

ldelossa replied:
These are terms used in AMQP spec. https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf

Reviewing this will clear things up. An exchange routes messages to a queue given some rules.

Allda commented on notifier/amqp/deliverer.go:
```diff
@@ -0,0 +1,77 @@
+	}
+	err = ch.Publish(
+		d.conf.Exchange.Name,
+		d.conf.RoutingKey,
```
What is routing key?

ldelossa replied:
This is a term in the AMQP spec: https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf

A routing key tells an exchange which queue the message is destined for in the usual case.

Allda commented on notifier/postgres/deletenotifications.go:
```diff
@@ -0,0 +1,59 @@
+
+// deleteNotifications garbage collects notifications and their associated
+// id and receipt rows
+func deleteNotifications(ctx context.Context, pool *pgxpool.Pool, id uuid.UUID) error {
```
Do you really need to delete notifications from the database? Isn't it better to mark it as processed and keep it in DB for future debugging in case something went wrong?

ldelossa replied:
If a notification is not delivered, it is marked undelivered and notifier will retry. 

After we successfully deliver however, we need to GC at some point, or we will always grow the DB unbounded.

Allda reviewed (commented):
I made a very quick review of changes mainly focused on amqp part. I added a few comments there because I don't think it supports what we agree on in: https://projects.engineering.redhat.com/browse/CLOUDWF-136

I'll try to run it locally to see a real notification traffic.

hdonnay commented on notifier/postgres/setdeleted.go:
```diff
@@ -0,0 +1,28 @@
+// notification id
+func setDeleted(ctx context.Context, pool *pgxpool.Pool, id uuid.UUID) error {
+	const (
+		query = `UPDATE receipt SET status = 'deleted', ts = $1 WHERE notification_id = $2`
```
If you use the database's clock when creating the receipt row, you should use it when updating the row:
```suggestion
		query = `UPDATE receipt SET status = 'deleted', ts = CURRENT_TIMESTAMP WHERE notification_id = $1`
```

ldelossa replied:
using db timestamps now.

hdonnay commented on notifier/postgres/setdelivered.go:
```diff
@@ -0,0 +1,28 @@
+// notification id
+func setDelivered(ctx context.Context, pool *pgxpool.Pool, id uuid.UUID) error {
+	const (
+		query = `UPDATE receipt SET status = 'delivered', ts = $1 WHERE notification_id = $2`
```
Same comment about using the database's clock.

hdonnay commented on notifier/postgres/setdeliveryfailed.go:
```diff
@@ -0,0 +1,28 @@
+// notification id
+func setDeliveryFailed(ctx context.Context, pool *pgxpool.Pool, id uuid.UUID) error {
+	const (
+		query = `UPDATE receipt SET status = 'delivery_failed', ts = $1 WHERE notification_id = $2`
```
database clock

hdonnay commented on notifier/postgres/setdeliveryfailed.go:
```diff
@@ -0,0 +1,28 @@
+
+// setDeliveryFailed will update the receipt's status to "delivery_failed" for the provied
+// notification id
+func setDeliveryFailed(ctx context.Context, pool *pgxpool.Pool, id uuid.UUID) error {
```
All these functions that update the receipt table should just go in one file. They all take the same parameter(s) with a variable query and handle the results the same way.

ldelossa replied:
I'm usually a proponent of a function per file, makes finding things easier.

hdonnay commented on notifier/postgres/teststore.go:
```diff
@@ -0,0 +1,65 @@
+	DefaultDSN = `host=localhost port=5432 user=clair dbname=clair sslmode=disable`
+)
+
+func TestStore(ctx context.Context, t testing.TB) (*sqlx.DB, *Store, func()) {
```
I think this would be better if written to just return `(*pgxpool.Pool, func())`, because the caller can then either make a store using the pool or use it to examine the database directly or, as is usual in tests, both. It worked well for claircore's vulnstore with minimal changes.

ldelossa replied:
I think I need this for distlocks

hdonnay commented on pkg/pager/pager.go:
```diff
@@ -0,0 +1,60 @@
+//
+// The offset and limit output arguments are to be used in a sql query to
+// retrieve the correct subset of elements for the requested page.
+func SQLPager(inPage Page, count uint64) (outPage Page, offset uint64, limit uint64) {
```
`pager.SQLPager` is a clunky name

ldelossa replied:
Think this may go away with seeking pagination.

ldelossa replied:
this is obsolete now.

hdonnay commented on pkg/pager/pager.go:
```diff
@@ -0,0 +1,60 @@
+package pager
```
Why is this a separate package? If this is part of the notifier API, it should just live in that package.

ldelossa replied:
The idea was to eventually move this to ClairCore and share it, however this may go away completely with a seeking pager method

hdonnay replied:
Yeah, there's only a single struct in here, it should just get put in the `notifier` package.

hdonnay commented on pkg/pager/pager.go:
```diff
@@ -0,0 +1,60 @@
+	pages = math.Ceil(pages)
+	if pages <= 0 {
+		pages++
+	}
```
Isn't this just:

```
pages := count/size
if count % size != 0 {
    pages++
}
if pages == 0 {
    pages++
}
```

ldelossa replied:
again, probably gonna go away with seek pager implementation

ldelossa replied:
this is obsolete now.

hdonnay commented on pkg/pager/pager.go:
```diff
@@ -0,0 +1,60 @@
+	// the max number of elements returned in a page
+	Size uint64
+	Next int64
+}
```
having `Next` instead be something like `Cursor uuid.UUID` would allow non-offset based paging.

ldelossa replied:
done.

hdonnay commented on notifier/postgres/notifications.go:
```diff
@@ -0,0 +1,63 @@
+func notifications(ctx context.Context, pool *pgxpool.Pool, id uuid.UUID, page *pager.Page) ([]notifier.Notification, pager.Page, error) {
+	const (
+		query      = "SELECT id, body FROM notification_body WHERE notification_id = $1"
+		pagedQuery = "SELECT id, body FROM notification_body WHERE notification_id = $1 OFFSET $2 LIMIT $3"
```
This query's response time will get worse the further "back" it pages. Both `pagedQuery` and `query` *always* return the same number of rows from the execution engine. `PagedQuery` just then has the server do a linear read to the offset row, then return a limited number of them.

```suggestion
		pagedQuery = "SELECT id, body FROM notification_body WHERE notification_id = $1 AND id < $2 ORDER BY id desc LIMIT $3"
```
This, along with making the Page structure record the id of the last item in the page should solve that. Solution untested, though.

ldelossa replied:
fixed.

hdonnay commented on notifier/postgres/failed.go:
```diff
@@ -0,0 +1,30 @@
+	clairerror "github.com/quay/clair/v4/clair-error"
+)
+
+// created will return all notification ids in "created" status
```
Doc copy-paste. This logic should probably be combined with the `created` and `deleted` functions and just pass in the status to look for.

ldelossa replied:
fixed

hdonnay commented on notifier/postgres/deleted.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+// deleted will return all notification ids in "deleted" status
+func deleted(ctx context.Context, pool *pgxpool.Pool) ([]uuid.UUID, error) {
```
See `failed` comment.

hdonnay commented on notifier/postgres/created.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+// created will return all notification ids in "created" status
+func created(ctx context.Context, pool *pgxpool.Pool) ([]uuid.UUID, error) {
```
See `failed` comment.

hdonnay commented on notifier/postgres/deletenotifications.go:
```diff
@@ -0,0 +1,59 @@
+		Str("component", "notifier/postgres/deleteNotification").
+		Logger()
+
+	const (
+		deleteNotificationID = `DELETE FROM notification WHERE id = $1`
+		deleteNotifications  = `DELETE FROM notification_body WHERE notification_id = $1`
+		deleteReceipt        = `DELETE FROM receipt WHERE notification_id = $1`
+	)
```
Can't these cascade off the notification row?

hdonnay commented on notifier/store.go:
```diff
@@ -0,0 +1,85 @@
+	Deleted(ctx context.Context) ([]uuid.UUID, error)
+	// SetDelivered marks the provided notification id as delivered
+	SetDelivered(ctx context.Context, id uuid.UUID) error
+	// SetDeliveryFailed marks the provided notification id failed to be delivered
+	SetDeliveryFailed(ctx context.Context, id uuid.UUID) error
```
```suggestion
	// SetFailed marks the provided notification id failed to be delivered.
	SetFailed(ctx context.Context, id uuid.UUID) error
```

ldelossa replied:
If you look a little closer, these method names are an echo of the enum names in the db. This method literally means "set the delivery_failed enumeration".

hdonnay commented on notifier/store.go:
```diff
@@ -0,0 +1,85 @@
+	// PutNotifications must create a Receipt with status created status on
+	// successful persistence of notifications in such a way that Receipter.Created()
+	// returns the persisted notification id.
+	PutNotifications(ctx context.Context, opts PutOpts) error
```
This struct should just be arguments.

Put may be a weird verb to use here, it suggests HTTP `PUT` (like, updatable) to me, but this is a create/store method.

ldelossa replied:
No need to bloat the function argument list, a struct is common way to hold a grouping of arguments to a method of function.

hdonnay commented on notifier/postgres/putnotifications.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	// update known update operations
+	_, err = tx.Exec(ctx, insertUpdateOperation, opts.Updater, opts.UpdateID, time.Now())
```
Why is the tag only sometimes checked?

ldelossa replied:
fixed

hdonnay commented on notifier/vulnsummary.go:
```diff
@@ -0,0 +1,25 @@
+	Severity     string                  `json:"severity"`
+}
+
+func (vs *VulnSummary) FromVulnerability(v claircore.Vulnerability) {
```
This should take a pointer, because we pass Vulnerabilities as pointers everywhere.

ldelossa replied:
I don't want this function to modify the incoming vuln, why pass a pointer? If the calling code has a pointer, It can deref it.

hdonnay replied:
Because this forces a copy of the struct, just to copy everything out of it.

hdonnay commented on notifier/migrations/migration1.go:
```diff
@@ -0,0 +1,39 @@
+	--- a relation mapping notifications to their serialized json bodies
+	CREATE TABLE IF NOT EXISTS notification_body (
+		id uuid PRIMARY KEY,
+		notification_id uuid REFERENCES notification,
```
Cascade on delete?

hdonnay commented on notifier/migrations/migration1.go:
```diff
@@ -0,0 +1,39 @@
+	--- a relation expressing the current status of a notification
+	--- this acts as a trigger for application business logic
+	CREATE TABLE IF NOT EXISTS receipt (
+		notification_id uuid PRIMARY KEY REFERENCES notification,
```
Cascade on delete?

hdonnay commented on notifier/migrations/migration1.go:
```diff
@@ -0,0 +1,39 @@
+	--- this acts as a trigger for application business logic
+	CREATE TABLE IF NOT EXISTS receipt (
+		notification_id uuid PRIMARY KEY REFERENCES notification,
+		uo_id uuid REFERENCES notifier_update_operation(uo_id),
```
If the notifications could be made to cascade on deletion, this could be handled by a trigger that deletes the update operation row if nothing in this table points references it.

hdonnay commented on notifier/processor.go:
```diff
@@ -0,0 +1,231 @@
+// process is intended to be ran as a go routine.
+//
+// implements a blocking event loop via a time.Ticker
+func (p *Processor) process(ctx context.Context, c <-chan Event) {
```
Don't have a function immediately just `go anotherfuction()`. The caller can just put `go` in front of this method call if they want, but this removes the choice.

ldelossa replied:
If the purpose of the function is to be asynchronous, and the usage is explained, removing the need for the caller to write more code is not a bad thing, the intention is for these structs to be asynchronous processors.

hdonnay commented on notifier/processor.go:
```diff
@@ -0,0 +1,231 @@
+
+// process is intended to be ran as a go routine.
+//
+// implements a blocking event loop via a time.Ticker
```
`time.Ticker`?

ldelossa replied:
fixed

hdonnay commented on notifier/processor.go:
```diff
@@ -0,0 +1,231 @@
+		log.Info().Str("new", current.Ref.String()).Msg("newer update operation is present, will not process notifications")
+		return false
+	}
+	p.previous = prev.Ref
```
Doing this return via member assignment is needlessly complicated.

ldelossa replied:
the safe method is designed to just return a bool on whether or not the processor should continue processing.

hdonnay replied:
OK, but it actually returns `(ok, previousRef)`, it just does the second one via assigning it to shared state instead of returning it.

Having this method return it and the `create` method take it would make the sequence dependency explicit.

ldelossa replied:
I changed it.

hdonnay commented on notifier/processor.go:
```diff
@@ -0,0 +1,231 @@
+			n := Notification{
+				Manifest:      digest,
+				Reason:        r,
+				Vulnerability: vs,
```
Not creating a separate VulnSummary struct would avoid a copy here.

ldelossa replied:
good catch, updating to:

```
			// summarize most severe vuln affecting manifest
			// the vulns array will be sorted by most severe
			vuln := affected.Vulnerabilities[vulns[0]]

			digest, err := claircore.ParseDigest(manifest)
			if err != nil {
				return err
			}
			n := Notification{
				Manifest: digest,
				Reason:   r,
			}
			n.Vulnerability.FromVulnerability(*vuln)

			notifications = append(notifications, n)
```

hdonnay commented on notifier/poller.go:
```diff
@@ -0,0 +1,130 @@
+// Poll is a non blocking call which begins
+// polling the Matcher for UpdateOperations.
+//
+// Returned channel can be listened too for events.
```
```suggestion
// Returned channel can be listened to for events.
```

hdonnay commented on notifier/poller.go:
```diff
@@ -0,0 +1,130 @@
+		latest := uo[0]
+		// confirm notifications were never created for this UOID.
+		var errNoReceipt clairerror.ErrNoReceipt
+		_, err := p.store.ReceiptByUOID(ctx, latest.Ref)
```
Isn't this check also done by the `Processor`?

ldelossa replied:
It's the same idea of "check, then check under lock". There's a fast pass check here, then the processor confirms the work still needs to be done and another worker did not complete it while it was parked or waiting.

hdonnay commented on notifier/delivery.go:
```diff
@@ -0,0 +1,169 @@
+		Str("deliverer", d.Deliverer.Name()).
+		Str("component", "notifier/delivery/Delivery.Deliver").Logger()
+	log.Info().Msg("delivering notifications")
+	go d.deliver(ctx)
+}
+
+// deliver is intended to be ran as a go routine.
+//
+// implements a blocking event loop via a time.Ticker
+func (d *Delivery) deliver(ctx context.Context) error {
```
Don't immediately just call `go` on another function. The caller can just add the go statement, but can't take it away.

ldelossa replied:
Calling go here removes the littering of go routine calls in the calling code. the usage is explained, and this provides a cleaner api to the caller.

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,88 @@
+	conf Config
+	// a client to use for POSTing webhooks
+	c   *http.Client
+	log *zerolog.Logger
```
`Log` member seems unused.

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,88 @@
+		Str("component", "notifier/webhook/deliverer.Deliver").
+		Logger()
+
+	callback := d.conf.callback
```
This isn't accessed under the lock.

ldelossa replied:
Deliverer now gets a copy of a Validated config. no locks necessary.

hdonnay commented on notifier/webhook/config.go:
```diff
@@ -0,0 +1,41 @@
+
+// Validate confirms configuration is valid and fills in private members
+// with parsed values on success.
+func (c *Config) Validate() error {
```
Why are only some the of the private members updated under lock? The private members also don't seem to be accessed under the lock.

ldelossa replied:
Parse is called by goroutines sharing a config, the fields which are written to are locked.

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,88 @@
+	if err != nil {
+		return err
+	}
+	buf := bytes.NewBuffer(b)
```
```suggestion
	buf := bytes.NewReader(b)
```

ldelossa replied:
done.

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,88 @@
+		Msg("dispatching webhook")
+
+	resp, err := d.c.Do(req)
+	if err != nil {
```
```suggestion
	if resp != nil {
		defer resp.Body.Close()
	}
	if err != nil {
```

hdonnay commented on notifier/amqp/config.go:
```diff
@@ -0,0 +1,125 @@
+	}
+	if c.TLSRootCA != "" {
+		TLS := tls.Config{}
+		TLS.RootCAs = x509.NewCertPool()
```
```suggestion
		TLS.RootCAs = x509.SystemCertPool()
```

This would make adding a CA to the system store work as expected, and make the CA argument optional.

ldelossa replied:
this will use the default cert repo on a linux system with common root ca's?

ldelossa replied:
fixed

hdonnay commented on notifier/amqp/failover.go:
```diff
@@ -0,0 +1,94 @@
+	f.RLock()
+	if f.conn != nil && !f.conn.IsClosed() {
+		log.Debug().Msg("existing connection exist and is not closed. returning this connection")
+		return f.conn, nil
```
This return leaves a read lock dangling.

hdonnay commented on notifier/amqp/config.go:
```diff
@@ -0,0 +1,125 @@
+
+// Config provides configuration for an AMQP deliverer.
+type Config struct {
+	sync.Mutex
```
A `FailOver` and a `Deliverer` can't share a config because of this lock, is that desirable?

ldelossa replied:
All configs have locks removed now, Validate() method returns a copy.

hdonnay commented on notifier/amqp/deliverer.go:
```diff
@@ -0,0 +1,78 @@
+		return &clairerror.ErrDeliveryFailed{err}
+	}
+
+	callback := d.conf.callback
```
Not accessed under lock

ldelossa replied:
Remedied by Validate() methods returning a copy, locks are gone.

hdonnay commented on notifier/amqp/failover.go:
```diff
@@ -0,0 +1,94 @@
+			log.Info().Str("broker", uri).Msg("failed to connect to AMQP broker. attempting next broker")
+			continue
+		}
+		ch, err := conn.Channel()
```
The `ch.Close()` method is only called on the success path. Is that OK?

ldelossa replied:
Think so, since every error path calls "conn.Close()" and their docs say:

"After returning from this call, all resources associated with this connection,
including the underlying io, Channels, Notify listeners and Channel consumers
will also be closed."

hdonnay commented on config/config.go:
```diff
@@ -51,190 +48,17 @@ type Config struct {
+	LogLevel string  `yaml:"log_level" json:"log_level"`
+	Indexer  Indexer `yaml:"indexer" json:"indexer"`
+	Matcher  Matcher `yaml:"matcher"`
+	Notifier Notifier
```
Missing tag

ldelossa replied:
fixed

hdonnay commented on config/introspection.go:
```diff
@@ -0,0 +1,35 @@
+	ServiceName string            `yaml:"service_name" json:"service_name"`
+	Tags        map[string]string `yaml:"tags" json:"tags"`
+	BufferMax   int               `yaml:"buffer_max" json:"buffer_max"`
+}
```
This changes the config by removing the `inline` annotation to the struct tags.

The newly un-inlined keys are now `Capped` instead of `lower`.

The configuration options now stutter: `Agent: {agent_endpoint: ""}`.

These changes shouldn't be made in this commit, if at all.

ldelossa replied:
A subset of these changes are already in master, these changes came in to remedy a rebase.

hdonnay commented on notifier/service/service.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+// Opts configures the notifier service
+type Opts struct {
```
This is just a re-statement of `config.Notifier`, why is it needed?

hdonnay commented on notifier/service/service.go:
```diff
@@ -0,0 +1,223 @@
+	"github.com/quay/clair/v4/indexer"
+	"github.com/quay/clair/v4/matcher"
+	"github.com/quay/clair/v4/notifier"
+	namqp "github.com/quay/clair/v4/notifier/amqp"
```
Why is this renamed?

ldelossa replied:
I've been using samqp for steadway's and namqp for notifier's

hdonnay commented on notifier/service/service.go:
```diff
@@ -0,0 +1,223 @@
+	}
+
+	fo := &namqp.FailOver{
+		Config: *opts.AMQP,
```
This has an embedded lock, it's not okay to copy around.

ldelossa replied:
Locks are gone.

hdonnay commented on notifier/service/service.go:
```diff
@@ -0,0 +1,223 @@
+	ds := make([]*notifier.Delivery, 0, deliveries)
+	for i := 0; i < deliveries; i++ {
+		distLock := pgdl.NewLock(lockPool, 0)
+		wh, err := webhook.New(*opts.Webhook, nil)
```
This has an embedded lock, it's not OK to copy.

ldelossa replied:
locks are gone.

hdonnay commented on httptransport/client/matcher.go:
```diff
@@ -10,13 +11,39 @@ import (
+	}
+	buf := bytes.NewBuffer(b)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), buf)
+	if err != nil {
+		return nil, err
+	}
```
`Req.Body` isn't closed, status code isn't checked.

ldelossa replied:
fixed

hdonnay commented on httptransport/notificationshandler.go:
```diff
@@ -0,0 +1,160 @@
+		if err != nil {
+			resp := &je.Response{
+				Code:    "bad-request",
+				Message: "could not parse \"page\" query param into uuid",
```
uuid?

ldelossa replied:
now valid, since "next" should be a uuid.

```
		n, err := uuid.Parse(param)
		if err != nil {
			resp := &je.Response{
				Code:    "bad-request",
				Message: "could not parse \"next\" query param into uuid",
			}
			je.Error(w, resp, http.StatusBadRequest)
			return
		}
```

hdonnay commented on httptransport/notificationshandler.go:
```diff
@@ -0,0 +1,160 @@
+		Size: pageSize,
+		Next: page,
+	}
+	notifications, outP, err := h.serv.Notifications(ctx, notificationID, inP)
```
Does this mean it's impossible to call without paging?

ldelossa replied:
Yes, the API assumes paging.

hdonnay commented on httptransport/server.go:
```diff
@@ -6,6 +6,7 @@ import (
 	"net"
 	"net/http"
 
+	notifier "github.com/quay/clair/v4/notifier/service"
```
group with the local imports

hdonnay commented on httptransport/server.go:
```diff
@@ -300,7 +336,7 @@ func unmodified(r *http.Request, v string) bool {
 }
 
-// WriterError is a helper that closes over an error that may be returned after
+// writerError is a helper that closes over an error that may be returned after
```
Why un-capitalize the start of a sentence?

ldelossa replied:
the comment is referring to a private function, caps have meaning.

hdonnay commented on initialize/services.go:
```diff
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"time"
 
+	notifier "github.com/quay/clair/v4/notifier/service"
```
Group with the local imports

hdonnay commented on initialize/services.go:
```diff
@@ -42,8 +43,39 @@ func (i *Init) Services() error {
 		}
+
+		// configure notifier service
+		dInterval, err := time.ParseDuration(i.conf.Notifier.DeliveryInterval)
+		if err != nil {
+			return &clairerror.ErrNotInitialized{
+				Msg: "notifier: failed to parse delivery interval: " + err.Error(),
+			}
+		}
+		pInterval, err := time.ParseDuration(i.conf.Notifier.PollInterval)
+		if err != nil {
+			return &clairerror.ErrNotInitialized{
+				Msg: "notifier: failed to parse poll interval: " + err.Error(),
+			}
+		}
+
```
Ah, I see, the config and the notifier opts have some differing types.

Why doesn't the config just have the needed types? The yaml package handles Durations correctly: https://play.golang.org/p/MyhYUozoQBo

ldelossa replied:
Just following the current pattern of things.

Allda commented on httptransport/server.go:
```diff
@@ -25,6 +26,7 @@ const (
 	IndexAPIPath            = apiRoot + "index_report"
 	IndexReportAPIPath      = apiRoot + "index_report/"
 	IndexStateAPIPath       = apiRoot + "index_state"
+	NotificationAPIPath     = apiRoot + "notification/"
```
OpenAPI spec defines it as plural ``/api/v1/notifications/

ldelossa replied:
fixed

Allda commented on initialize/services.go:
```diff
@@ -42,8 +43,39 @@ func (i *Init) Services() error {
+			Matcher:          libV,
+			Migrations:       i.conf.Notifier.Migrations,
+			PollInterval:     pInterval,
+			Webhook:          i.conf.Notifier.Webbook,
```
AMQP is missing in Combo mode

ldelossa replied:
fixed, great catch.

hdonnay commented on config/config.go:
```diff
@@ -247,6 +71,9 @@ func Validate(conf Config) error {
 		if conf.Matcher.ConnString == "" {
 			return fmt.Errorf("matcher mode requires a database connection string")
 		}
+		if conf.Notifier.ConnString == "" {
+			return fmt.Errorf("notifier mode requires a database connection string")
+		}
```
```suggestion
		if conf.Notifier.ConnString == "" {
			conf.Notifier.ConnString = conf.Matcher.ConnString
		}
```

ldelossa replied:
This would be odd one-off thing to do for notifier. Right now the config requires a db string be present for each service so that it works the same if you decide not to run in combo-mode.

I do see what you're getting at here, we only need to "find" one db string for combo mode, but in combo mode can also be pointed to different databases, its probably better to be explicit in regards to database strings since it works the same for any mode.

hdonnay replied:
I forgot the notifier has its own DB; this suggestion sucks.

hdonnay commented on config/introspection.go:
```diff
@@ -0,0 +1,35 @@
+type Jaeger struct {
+	Agent struct {
+		Endpoint string `yaml:"endpoint" json:"endpoint"`
+	} `yaml:"agent"`
```
Needs json tags if we're changing this because of json.

hdonnay commented on httptransport/client/matcher.go:
```diff
@@ -10,13 +11,50 @@ import (
+	}
+	if resp.Body != nil {
+		defer resp.Body.Close()
+	}
```
```suggestion
	if resp.Body != nil {
		defer resp.Body.Close()
	}
	if err != nil {
		return nil, err
	}
```

hdonnay commented on httptransport/notificationshandler.go:
```diff
@@ -0,0 +1,158 @@
+	"github.com/quay/clair/v4/notifier/service"
+	"github.com/quay/clair/v4/pkg/pager"
+	je "github.com/quay/claircore/pkg/jsonerr"
+	"github.com/rs/zerolog"
```
```suggestion
	"github.com/google/uuid"
	je "github.com/quay/claircore/pkg/jsonerr"
	"github.com/rs/zerolog"

	"github.com/quay/clair/v4/notifier"
	"github.com/quay/clair/v4/notifier/service"
	"github.com/quay/clair/v4/pkg/pager"
```

hdonnay commented on notifier/amqp/deliverer_integration_test.go:
```diff
@@ -0,0 +1,103 @@
+	"os"
+	"testing"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/google/uuid"
+	"github.com/quay/claircore/test/integration"
+	samqp "github.com/streadway/amqp"
```
```suggestion
	"golang.org/x/sync/errgroup"
	"github.com/google/uuid"
	"github.com/quay/claircore/test/integration"
	samqp "github.com/streadway/amqp"
```

hdonnay commented on notifier/service/service.go:
```diff
@@ -0,0 +1,276 @@
+// Service is an interface wrapping ClairV4's notifier functionality.
+type Service interface {
+	// Retrieves an optional paginated set of notifications given an notification id
+	Notifications(ctx context.Context, id uuid.UUID, page *pager.Page) ([]notifier.Notification, pager.Page, error)
```
`Pager.Page` should just be a part of the `notifier` package.

ldelossa replied:
That seems not right, this struct will be used for *any* paging api we use moving forward. the chances of it being shared and being moved to another package is high.

hdonnay commented on notifier/stomp/config.go:
```diff
@@ -0,0 +1,97 @@
+	Passcode string `yaml:"passcode"`
+}
+
+type Config struct {
```
Yaml tags are inconsistent on this struct.

Allda commented on config/config.go:
```diff
@@ -51,190 +48,17 @@ type Config struct {
+	LogLevel string   `yaml:"log_level" json:"log_level"`
+	Indexer  Indexer  `yaml:"indexer" json:"indexer"`
+	Matcher  Matcher  `yaml:"matcher"`
+	Notifier Notifier `yaml:"notifer"`
```
s/notifer/notifier/

Allda commented on notifier/processor.go:
```diff
@@ -0,0 +1,231 @@
+
+	uos := all[e.updater]
+	n := len(uos)
+	if n < 2 {
```
Are you going to change it to produce notification for dynamic updaters?

hdonnay replied:
Yeah, this needs to return `true, uuid.NIl` and the decision to not send needs to be made once the diff size is seen.

ldelossa replied:
@hdonnay @Allda can we talk about this change a bit?

Ideally, we do not want to perform the request for a very large diff, as we'd be asking the matcher to do all the work and return a large response just to say we won't process the notification.

Ideally this change happens at the matcher, and the matcher can return a `413 Payload Too Large` saying the diff is too large to process. However this still has the matcher do a least a "count" query. Maybe this is not a bad thing? 

You see that the safest thing to do is simply not process an update diff if it's the only one. I'm not held doing this, but lets discuss best way to do it moving forward.

Allda commented on notifier/stomp/deliverer.go:
```diff
@@ -0,0 +1,62 @@
+	if c, err = conf.Validate(); err != nil {
+		return nil, err
+	}
+	return &Deliverer{
```
Configuration is stored twice in Deliver and in Failover and both have separate TLS configurations. Why is that?

ldelossa replied:
They both use the same config. Validate() returns a copy of a Config with all private members parsed

Allda commented on notifier/stomp/failover.go:
```diff
@@ -0,0 +1,83 @@
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	_, err := f.Config.Validate()
```
This doesn't create a tls object in a config since the config is not a pointer

ldelossa replied:
what do you mean?

ldelossa replied:
Validate() returns a copy of the config, in this case we were only confirming that the config is still valid.

Allda commented on notifier/stomp/failover.go:
```diff
@@ -0,0 +1,83 @@
+
+	var conn io.ReadWriteCloser
+	var err error
+	if f.tls != nil {
```
In integration tests this is always nil, because `Validate` doesn't return updated config with tls setup

Allda commented on notifier/stomp/deliverer.go:
```diff
@@ -0,0 +1,62 @@
+func New(conf Config, fo *FailOver) (*Deliverer, error) {
+	var c Config
+	var err error
+	if c, err = conf.Validate(); err != nil {
```
`Validate()` is actually called twice and tls is also configured twice in `Deliverer` and in `Failover`. Is that really needed?

ldelossa replied:
I think you're right, I'll make failOver private and have the Deliverer constructor create it.

Commit:
implement keyserver

this commit implements a keystore for persisting public keys, a
keymanager for managing public and private key stores, and the necessary
api plubming.

Signed-off-by: ldelossa <ldelossa@redhat.com>

hdonnay commented on httptransport/keybyidhandler.go:
```diff
@@ -0,0 +1,77 @@
+	"net/http"
+	"path"
+
+	jose "gopkg.in/square/go-jose.v2"
+
+	"github.com/google/uuid"
+	clairerror "github.com/quay/clair/v4/clair-error"
+	"github.com/quay/clair/v4/notifier"
+	je "github.com/quay/claircore/pkg/jsonerr"
```
```suggestion
	jose "gopkg.in/square/go-jose.v2"
	je "github.com/quay/claircore/pkg/jsonerr"
	"github.com/google/uuid"

	clairerror "github.com/quay/clair/v4/clair-error"
	"github.com/quay/clair/v4/notifier"
```

ldelossa replied:
@hdonnay this is the default gofmt ordering, if I change the ordering, my editor snaps it right back. Also any external contributions will not perform this ordering. 

if gofmt support a config file that we can drop at root, I would say we can enforce this, but for now I won't enforce.

hdonnay commented on httptransport/keybyidhandler.go:
```diff
@@ -0,0 +1,77 @@
+			}
+			je.Error(w, resp, http.StatusInternalServerError)
+			return
+
```
```suggestion
```

ldelossa replied:
my source file does not have this space, may have been formatted away. lets see if it's still there after this next round of fixes get pushed.

hdonnay commented on notifier/keymanager/manager.go:
```diff
@@ -0,0 +1,210 @@
+		Str("component", "notifier/keymanager/NewManager").
+		Logger()
+
+	var m *Manager = &Manager{
+		kp:    atomic.Value{},
+		store: store,
+	}
```
```suggestion
	m := &Manager{
		kp:    atomic.Value{},
		store: store,
	}
```

ldelossa replied:
any reason?

hdonnay replied:
It's the idiom. Writing the declaration out the long way makes me think something weird is happening, but I'm pretty sure there's not.

ldelossa replied:
I think this is fine, its valid syntax, is explicit about what it's doing, and not ambiguous.

hdonnay commented on notifier/keymanager/manager.go:
```diff
@@ -0,0 +1,210 @@
+		kp:    atomic.Value{},
+		store: store,
+	}
+	newKP, err := m.genKeyPair(ctx)
```
This shouldn't be a method because it doesn't mutate or reference anything from the Manager.

ldelossa replied:
I do not want this method at the package level, this is being used simply as a namespace.

hdonnay commented on notifier/keymanager/manager.go:
```diff
@@ -0,0 +1,210 @@
+	log.Info().Msg("gc starting")
+	var total int64
+	var err error
+	var n int64 = math.MaxInt64
```
Why is this initialized to MaxInt instead of just -1?

ldelossa replied:
fair. I think at first I had <=.

ldelossa replied:
fixed

hdonnay commented on notifier/keymanager/manager.go:
```diff
@@ -0,0 +1,210 @@
+		}
+		// atomic store
+		m.kp.Store(&newKP)
+		return nil
```
Why does this create a new key instead of just re-using the existing one?

ldelossa replied:
The key has expired, if we want to promote the rotation of keys after expiration, we should create a new one.

hdonnay commented on notifier/migrations/migration1.go:
```diff
@@ -0,0 +1,50 @@
+	CREATE TABLE IF NOT EXISTS key (
+	    id uuid PRIMARY KEY,
+		expiration timestamptz,
+		pub_key bytea
```
indentation messed up

ldelossa replied:
ran SQL thru a formatter.

hdonnay commented on notifier/migrations/migration1.go:
```diff
@@ -0,0 +1,50 @@
+		expiration timestamptz,
+		pub_key bytea
+	);
+	CREATE INDEX key_idx ON key (id);
```
I think this index is redundant because the primary key always has an index.

ldelossa replied:
fixed

hdonnay commented on notifier/keymanager/manager.go:
```diff
@@ -0,0 +1,204 @@
+	err := m.store.BumpExpiration(ctx, kp.ID, delta)
+	switch {
+	case errors.As(err, &clairerror.ErrKeyNotFound{}):
+		newKP, err := m.genKeyPair(ctx)
+		if err != nil {
+			return err
+		}
+		if err := m.store.PutKey(ctx, newKP.ID, newKP.Public, delta); err != nil {
+			return err
+		}
+		// atomic store
+		m.kp.Store(&newKP)
+		return nil
```
Why does this generate a new keypair, instead of just using the existing keypair with a new expiration?

ldelossa replied:
The key has expired, if we want to promote the rotation of keys after expiration, we should create a new one.

hdonnay commented on notifier/postgres/keystore.go:
```diff
@@ -0,0 +1,188 @@
+		ID:         tmp.id,
+		Expiration: tmp.exp,
+		Public:     pub,
+	}
```
Can't you just Scan directly into a `Key` struct and a byte slice? I don't think the extra struct buys anything for readability.

ldelossa replied:
fixed.

hdonnay commented on notifier/stomp/config.go:
```diff
@@ -0,0 +1,97 @@
+	Rollup int `yaml:"rollup"`
+	// The callback url where notifications are retrieved.
+	Callback string `yaml:"callback"`
+	callback url.URL
```
Why isn't this a pointer? I think it might be a holdover from the old locked version of this struct. Turning this into a pointer makes the code in `Validate` simpler.

ldelossa replied:
the url struct is rather small, why not remove the possibility of it winding up on the GC sweep by copying it around?

hdonnay commented on notifier/stomp/directdeliverer.go:
```diff
@@ -0,0 +1,99 @@
+		}
+
+		n := d.n[bs:be]
+		b, err := json.Marshal(&n)
```
This will generate a lot of separate slices. I think it will be easier on the gc to reuse a bytes.Buffer and json.Encoder for each iteration of the loop, e.g.:

```
buf := &bytes.Buffer{}
enc := json.NewEncoder(buf)
for ... {
    ...
    buf.Reset()
    if err := enc.Encode(&n); err != nil {
        ...
    }
    ...
    tx.Send(..., buf.Bytes(), ...)
}
```

This would result in just two big buffers getting used, rather than 1 + number of iterations.

Its a real shame there's not an API for the stdlib json that doesn't involve buffering the whole data structure internally.

ldelossa replied:
fixed

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,144 @@
+	expire := jwt.NumericDate(time.Now().Add(1 * time.Hour).Unix())
+	cl := jwt.Claims{
+		Subject:  "clair-notifier",
+		Issuer:   "clair-notifier",
```
I think this clashes with the hardcoded path in the notifier. This would result in requests for `/services/clair-notifier/keys/{id}`

ldelossa replied:
I'm not seeing how issuer field clashes with anything else in the app? Isn't the only thing concerned with this value is the receiver of the JWT. Does "issuer" get used for anything other then the receiving code's verification logic.

hdonnay replied:
because the service name is hardcoded to just "notifier" in the path, but this would cause the verifier to look up "clair-notifier"

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,144 @@
+		Expiry:   &expire,
+		IssuedAt: &now,
+	}
+	kIDClaim := struct {
```
This needs to go into the `(jwt.JsonWebToken).Headers`, not the claims. It's very possible I implemented the auth incorrectly over in that part of the code.

ldelossa replied:
fixed

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,144 @@
+	if err != nil {
+		return &clairerror.ErrDeliveryFailed{E: err}
+	}
+	if resp.StatusCode != http.StatusOK {
```
The server will return OK and not Created or Accepted?

ldelossa replied:
@hdonnay I have not worked with a webhook system where the server receiving a webhook does not respond with 200OK.

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,144 @@
+		return err
+	}
+	req.Header.Set("Authorization", "Bearer "+token)
+	req.Header.Set("kid", kp.ID.String())
```
The keyserver protocol doesn't mandate this as an HTTP header.

ldelossa replied:
fixed

hdonnay commented on notifier/webhook/deliverer_test.go:
```diff
@@ -0,0 +1,163 @@
+		t.Fatalf("failed to sign request: %v", err)
+	}
+
+	kid := req.Header.Get("kid")
```
The spec says nothing about this HTTP header

ldelossa replied:
fixed

hdonnay commented on notifier/webhook/deliverer.go:
```diff
@@ -0,0 +1,144 @@
+//
+// a kid header specifying the private key which signed the jwt is attached as well.
+func (d *Deliverer) sign(ctx context.Context, req *http.Request, kp keymanager.KeyPair) error {
+	signer, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.PS512, Key: kp.Private}, (&jose.SignerOptions{}).WithType("JWT"))
```
The keyid needs to be set here

ldelossa replied:
fixed

hdonnay reviewed (changes requested):
lgtm with the webhook jwt issuer changed to match the service name exported on the keyserver

Commit:
remove pager package

this commit removes pager package infavor of keeping the page structure
local to the notifier package.

Signed-off-by: ldelossa <ldelossa@redhat.com>

hdonnay reviewed (approved):
LGTM
